### DISCUSSION TITLE
How to repeat the same audio files that randomly selected in the first place

### DISCUSSION THREAD

**Discussion by pavlos2** [savonet/liquidsoap#2862] [Q&A]
hi ls users! I need help on how to play the same audio files more than once. The setup is as shown below and I need to have the two audio files from part 2 to play two more times, the same files that were randomly selected in the first place.

```
part1 = rotate (weights=[1,1], [playlist_01, playlist_02])
part2 = rotate (weights=[1,1], [playlist_03, playlist_04])
part3 = rotate(weights=[1,1], [playlist_05, playlist_06])
Joyful = rotate(weights=[2,6,2], [part1, part2, part3])

radio = switch(track_sensitive=true, [
([predicate.at](http://predicate.at/)_most(10, {00m-15m  or 30m-45m}), Joyful),
({true}, radio)])
```

Many thanks in advance!

**Comment by pavlos2** [comment]
Hi @smimram,

I see you edited my question by modifying the predicate operator, is it an answer to my question? Please help.

Regards,
Pavlos

**Reply by vitoyucepi** [reply]
Hi @pavlos2,
I've additionally reformatted your code, but I don't understand the question.
I think it needs some kind of diagram.

```ruby
part1 = rotate(weights=[1, 1], [playlist_01, playlist_02])
part2 = rotate(weights=[1, 1], [playlist_02, playlist_03])
part3 = rotate(weights=[1, 1], [playlist_04, playlist_05])
Joyful = rotate(weights=[2, 6, 2], [part1, part2, part3])

radio = switch(track_sensitive=true, [
  (predicate.at_most(10, {00m-15m  or 30m-45m}), Joyful),
  ({true}, radio)
])
```

Hmm, `predicate.at_most` is from the *development* version.

**Comment by pavlos2** [comment]
Hi Vito and thank you for the response. My question is mainly on how to handle the rotation called Joyful so that it will output:

2 randomly selected files from part1  -->  [audio1, audio2]
2 randomly selected files from part2 that will be repeated n times, **the same files** --> [audio3, audio4, audio3, audio4, audio3, audio4, ...]
2 randomly selected files from part3  -->  [audio5, audio6]

Finally, in the switch statement for the radio stream I use the predicate.at_most function to play only these n files every half an hour.

Thanks in advance for your help.

Regards,
Pavlos

**Reply by vitoyucepi** [reply]
I concluded that you should write your playlist function for that.
If I were you, I would use the code on [page 77](http://www.liquidsoap.info/book/book.pdf) as a base.

In the custom playlist function, I will load all the playlists and store a complete list of files for one full rotation.
Each time the `next()` is called, I will check and refill that rotation list and global playlists.

**Comment by pavlos2** [comment]
Thank you Vito for refreshing my question to more algorithm thinking. I tried the suggested code on page 77, to no avail. Can you help with the script please? I was hoping that I could use an operator for part2 (in the Joyfull playlist) to iterate through the randomly selected files n times.

**Reply by vitoyucepi** [reply]
Here you are
```ruby
def joyful_playlist()
  # Function to keep track of files
  def load_playlist_files(path)
    files = {
      # List of files in the playlist
      list = ref([]),
      # Playlist path
      path = path
    }

    # Function to load the next file
    let files.next = fun() -> begin
      # Refill if list is empty
      if list.is_empty(!(files.list)) then
        files.list := playlist.files(files.path)
      end

      # Get the first element of the list
      next_file_path = list.hd(!(files.list))
      files.list := list.tl(!(files.list))

      next_file_path
    end
    files
  end

  playlist_1 = load_playlist_files('playlist_1')
  playlist_2 = load_playlist_files('playlist_2')
  playlist_3 = load_playlist_files('playlist_3')
  playlist_4 = load_playlist_files('playlist_4')
  playlist_5 = load_playlist_files('playlist_5')
  playlist_6 = load_playlist_files('playlist_6')
  queue = ref([])

  def next()
    if list.is_empty(!queue) then
      playlist_3_next = playlist_3.next()
      playlist_4_next = playlist_4.next()
      queue := [
        playlist_1.next(),
        playlist_2.next(),
        playlist_3_next,
        playlist_4_next,
        playlist_3_next,
        playlist_4_next,
        playlist_3_next,
        playlist_4_next,
        playlist_5.next(),
        playlist_6.next()
      ]
    end
    next_file = list.hd(!queue)
    queue := list.tl(!queue)

    request.create(next_file)
  end
  
  request.dynamic(next)
end

joyful = joyful_playlist()
```

The playlist log
```
playlist 1 - track 1
playlist 2 - track 1
playlist 3 - track 1
playlist 4 - track 1
playlist 3 - track 1
playlist 4 - track 1
playlist 3 - track 1
playlist 4 - track 1
playlist 5 - track 1
playlist 6 - track 1
playlist 1 - track 2
playlist 2 - track 2
playlist 3 - track 2
playlist 4 - track 2
playlist 3 - track 2
playlist 4 - track 2
playlist 3 - track 2
playlist 4 - track 2
playlist 5 - track 2
playlist 6 - track 2
```

**Reply by pavlos2** [reply]
Hi Vito, thank you so much for the script! I use the AzuraCast software and I try to understand how the function `load_playlist_files(path)` will get the path of each playlist.

**Reply by pavlos2** [reply]
Hi Vito, eventually I had some time to test the script, thank you so much once more! How can I prevent the selection of the same file from playlist_3 and playlist_4?

**Reply by vitoyucepi** [reply]
Hi @pavlos2,

> How can I prevent the selection of the same file from playlist_3 and playlist_4?

Could you explain in detail?

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

Because some of the audio files may be found in both lists, playlist_3 and playlist_4, I would like to prevent playing the same file in the rotation list. Probably we can achieve that by comparing the filenames before refilling the rotation list?

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

I modified the "next" function as following, so that to fill in the queue with unique filenames. I need to compare the two filenames but excluding the last character. I guess this this logic expression **`while playlist_1_next[:-1] == playlist_2_next[:-1] do`** is not probably written for liquidsoap, can you help? 

```
def next()
    if list.is_empty(!queue) then
      playlist_1_next = playlist_1.next()
      playlist_2_next = playlist_2.next()

 while playlist_1_next[:-1] == playlist_2_next[:-1] do
      playlist_1_next = playlist_1.next()
      playlist_2_next = playlist_2.next()
    end

      queue := [playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next,
                playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next,
                playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next,
                playlist_1_next, playlist_2_next]
    end
    next_file = list.hd(!queue)
    queue := list.tl(!queue)

    request.create(next_file)
  end
```

**Reply by vitoyucepi** [reply]
Hi, do you have programming skills in other languages like python?
To remove the last character, you have to `string.sub(s, start=0, length=string.length(s)-1)` where `s` is string.

The rest of your question is not so simple for me.
Sometimes I wish there were python bindings for liquidsoap.

If you want to remove complete dupes from the playlists, then I'll recommend you to use `list.filter` and `list.mem`.

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

Thanks once more. I will try it in the script...  Totally agree about the python bindings for liquidsoap. Also, I would like to see better detailed documentation with more examples of Liquidsoap.

**Reply by vitoyucepi** [reply]
As a source of documentation with examples, I use
1. [The book](http://www.liquidsoap.info/book/book.pdf)
1. [Reference site section](https://www.liquidsoap.info/doc-2.1.3/language.html)
1. [Libs](https://github.com/savonet/liquidsoap/tree/main/src/libs)
1. [Tests](https://github.com/savonet/liquidsoap/tree/main/tests/language)

If you know some other language, you could design a standalone app and interact with it using http or stdout. You can even utilize json if you want to.

**Reply by pavlos2** [reply]
Hi @vitoyucepi ,

Indeed these are good sources of documentation, I keep using them for sometime, however I would rather to see more examples as well as an online place to test Liquidsoap code. Anyway, that's only my personal opinion.

Regarding my original code, I'm stuck again. Here again is the entire code I've come up with so far with your help:

```
def decade_playlist()
  # Function to keep track of files
  def load_playlist_files(path)
    files = {
      # List of files in the playlist
      list = ref([]),
      # Playlist path
      path = path
    }

    # Function to load the next file
    let files.next = fun() -> begin
      # Refill if list is empty
      if list.is_empty(!(files.list)) then
        files.list := playlist.files(files.path)
      end

      # Get the first element of the list
      next_file_path = list.hd(!(files.list))
      files.list := list.tl(!(files.list))

      next_file_path
    end
    files
  end

  playlist_1 = load_playlist_files('/var/azuracast/stations/test/playlists/playlist_00_hm_part1.m3u')
  playlist_2 = load_playlist_files('/var/azuracast/stations/test/playlists/playlist_00_hm_part2.m3u')
  queue = ref([])

  def next()
    if list.is_empty(!queue) then
      playlist_1_next = playlist_1.next()
      playlist_2_next = playlist_2.next()


    #Insert to the queue only unique filenames (different voices) 
    while string.sub(playlist_1_next, start=0, length=string.length(playlist_1_next)-10) == string.sub(playlist_2_next, start=0, length=string.length(playlist_2_next)-10) do
      playlist_1_next = playlist_1.next()
      playlist_2_next = playlist_2.next()
    end

      queue := [playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next]

 end
    next_file = list.hd(!queue)
    queue := list.tl(!queue)

    request.create(next_file)
  end
  
  request.dynamic(next)
end

# final playlist to be used
hmdecade = decade_playlist()
```
The issue I am encountering is that, although I made sure that the while loop inserts to the queue only unique filenames (different voices), the code is sometimes filling the queue with files of the same voice (not unique filenames). Probably I need to ensure that the files in the queue are being selected in a unique and random order, instead of always following a certain order. I am a bit lost at this point.

Thanks again for all your support!

**Reply by vitoyucepi** [reply]
Hi,
> random order

`list.shuffle` check code on [page 77](http://www.liquidsoap.info/book/book.pdf)

I think I've finally come to a solution for your problem.
You need a separate function to load these playlists.
At load time, they should filter their list against each other.
1. Playlist 1 loads. Playlist 2 is empty. The filter will pass for all.
2. Playlist 2 loads and filters all tracks against playlist 1.
3. Playlist 1 becomes empty, loads, and filters against playlist 2 initial state.
4. Playlist 2 becomes empty, loads, and filters against playlist 1 initial state.

There are some edge cases. For example, if one of the lists becomes empty after filtering.

**Reply by pavlos2** [reply]
can you help with the code? It's getting a bit complicate...

**Reply by pavlos2** [reply]
I tried the following, so that each time the playlist files are refilled, they are first shuffled using list.shuffle. This ensures that the next file selected from each playlist is randomly chosen each time. The issue is not resolved...

 ```
   # Function to load the next file
    let files.next = fun() -> begin
      # Refill if list is empty
      if list.is_empty(!(files.list)) then
        files.list := playlist.files(files.path)
        files.list := list.shuffle(!(files.list))
      end

      # Get the first element of the list
      next_file_path = list.hd(!(files. List))
      files. List := list.tl(!(files.list))

      next_file_path
    end
    files
  end
```

**Reply by vitoyucepi** [reply]
Here you are
```ruby
def load_playlist_files(~randomize = false, path)
  id = string.id("load_playlist_files")
  let log.info = fun(~label=id, message) -> log.info(label=label, message)
  let log.debug = fun(~label=id, message) -> log.debug(label=label, message)

  files = {
    # Source id
    id = id,
    # Playlist path
    path = path,
    # List of files in the playlist
    list = ref([]),
    # List of files left in queue
    queue = ref([])
  }

  # Is queue empty
  let files.is_empty = fun() -> begin
    list.is_empty(!(files.queue))
  end

  # Load files to the playlist
  let files.reload = fun() -> begin
    log.info("Reload full playlist")
    files.list := playlist.files(files.path)
    log.debug("files.list=#{!(files.list)}")
  end
  
  let files.fill_queue = fun() -> begin
    log.info("Fill playlist queue")
    if randomize then
      files.queue := list.shuffle(!(files.list))
    else
      files.queue := !(files.list)
    end
    
    log.debug("randomize=#{randomize}")
    log.debug("files.list=#{!(files.list)}")
    log.debug("files.queue=#{!(files.queue)}")
  end

  # Get next track in the playlist
  let files.next = fun() -> begin
    log.info("Get next file in queue")
    if files.is_empty() then
      log.info("Queue is empty")
      null()
    else
      # Get the first element of the list
      next_file_path = list.hd(!(files.queue))
      files.queue := list.tl(!(files.queue))
      log.info("Next file in queue is: #{next_file_path}")
      log.debug("files.queue=#{!(files.queue)}")
      
      next_file_path
    end
  end

  files
end

# Generate rotation
def decade_playlist()

  playlist_1 = load_playlist_files(randomize = true, 'playlist_1.m3u')
  playlist_2 = load_playlist_files(randomize = true, 'playlist_2.m3u')
  queue = ref([])

  # Removes paths from playlist_2 if they also exist in playlist_1
  def remove_dupes(playlist_1, playlist_2)
    def is_duplicate(file_path)
      not list.mem(file_path, !(playlist_1.list))
    end
    playlist_2.list := list.filter(is_duplicate, !(playlist_2.list))
  end

  # Get next file in queue
  def next()
    if playlist_1.is_empty() then
      playlist_1.reload()
      remove_dupes(playlist_2, playlist_1)
      playlist_1.fill_queue()
    end
    if playlist_2.is_empty() then
      playlist_2.reload()
      remove_dupes(playlist_1, playlist_2)
      playlist_2.fill_queue()
    end
    if list.is_empty(!queue) then
      playlist_1_next = playlist_1.next()
      playlist_2_next = playlist_2.next()
      queue := [playlist_1_next, playlist_2_next, playlist_1_next, playlist_2_next]
    end

    next_file_path = list.hd(!queue)
    queue := list.tl(!queue)

    if null.defined(next_file_path) then
      request.create(null.get(next_file_path))
    else
      null()
    end
  end
  
  request.dynamic(next)
end

# final playlist to be used
hmdecade = decade_playlist()

```

**Reply by vitoyucepi** [reply]
I have playlists with duplicates
- `playlist_1.m3u`
  ```m3u
  playlist_1/track_1.ogg
  playlist_1/track_2.ogg
  playlist_1/track_3.ogg
  playlist_1/track_4.ogg
  playlist_2/track_1.ogg
  ```
- `playlist_2.m3u`
  ```m3u
  playlist_1/track_4.ogg
  playlist_2/track_1.ogg
  playlist_2/track_2.ogg
  playlist_2/track_3.ogg
  playlist_2/track_4.ogg
  ```

And here's my log

<details>
<summary>Details</summary>

```
./playlist_2/track_1.ogg
./playlist_2/track_4.ogg
./playlist_2/track_1.ogg
./playlist_2/track_4.ogg

./playlist_1/track_4.ogg
./playlist_2/track_2.ogg
./playlist_1/track_4.ogg
./playlist_2/track_2.ogg

./playlist_1/track_3.ogg
./playlist_2/track_3.ogg
./playlist_1/track_3.ogg
./playlist_2/track_3.ogg

./playlist_1/track_1.ogg
./playlist_2/track_2.ogg
./playlist_1/track_1.ogg
./playlist_2/track_2.ogg

./playlist_1/track_2.ogg
./playlist_2/track_4.ogg
./playlist_1/track_2.ogg
./playlist_2/track_4.ogg

./playlist_1/track_4.ogg
./playlist_2/track_3.ogg
./playlist_1/track_4.ogg
./playlist_2/track_3.ogg

./playlist_1/track_1.ogg
./playlist_2/track_4.ogg
./playlist_1/track_1.ogg
./playlist_2/track_4.ogg

./playlist_2/track_1.ogg
./playlist_2/track_2.ogg
./playlist_2/track_1.ogg
./playlist_2/track_2.ogg

./playlist_1/track_3.ogg
./playlist_2/track_3.ogg
./playlist_1/track_3.ogg
./playlist_2/track_3.ogg

./playlist_1/track_2.ogg
./playlist_2/track_4.ogg
./playlist_1/track_2.ogg
./playlist_2/track_4.ogg

./playlist_1/track_1.ogg
./playlist_2/track_3.ogg
./playlist_1/track_1.ogg
./playlist_2/track_3.ogg

./playlist_1/track_4.ogg
./playlist_2/track_2.ogg
./playlist_1/track_4.ogg
./playlist_2/track_2.ogg

./playlist_1/track_2.ogg
./playlist_2/track_4.ogg
./playlist_1/track_2.ogg
./playlist_2/track_4.ogg

./playlist_2/track_1.ogg
./playlist_2/track_3.ogg
./playlist_2/track_1.ogg
./playlist_2/track_3.ogg

./playlist_1/track_3.ogg
./playlist_2/track_2.ogg
./playlist_1/track_3.ogg
./playlist_2/track_2.ogg

./playlist_1/track_3.ogg
./playlist_2/track_4.ogg
./playlist_1/track_3.ogg
./playlist_2/track_4.ogg

./playlist_1/track_4.ogg
./playlist_2/track_2.ogg
./playlist_1/track_4.ogg
./playlist_2/track_2.ogg

./playlist_1/track_1.ogg
./playlist_2/track_3.ogg
./playlist_1/track_1.ogg
./playlist_2/track_3.ogg

./playlist_1/track_2.ogg
./playlist_2/track_2.ogg
./playlist_1/track_2.ogg
./playlist_2/track_2.ogg

./playlist_2/track_1.ogg
./playlist_2/track_4.ogg
./playlist_2/track_1.ogg
./playlist_2/track_4.ogg

./playlist_1/track_2.ogg
./playlist_2/track_3.ogg
./playlist_1/track_2.ogg
./playlist_2/track_3.ogg
```

</details>

As you can see *playlist_1* always rotates over 5 tracks and *playlist_2* has 3 tracks in it.

**Reply by pavlos2** [reply]
Hi @vitoyucepi, 

Thanks once more for your help, the script still does not resolve the issue, I believe it's the way we load the next file. Allow me to describe once more the main issue:

**I have two playlists, Playlist_1 and Playlist_2, and I need to rotate n times two randomly selected files, one from Playlist_1 and the other from Playlist_2. These two files must have a unique file name. When the rotation is complete, two new random files are chosen again to rotate n times and so on...**

Based on the code we have discussed so far I would start as follows:

```
def decade_playlist()
    # Function to keep track of files
      def load_playlist_files(~randomize = true, path)
        files = {
          # List of files in the playlist
          list = ref([]),
          # Playlist path
          path = path
        }
      end

      # Load the two playlists
      playlist_1 = load_playlist_files('/var/azuracast/stations/test/playlists/playlist_00_hm_part1.m3u')
      playlist_2 = load_playlist_files('/var/azuracast/stations/test/playlists/playlist_00_hm_part2.m3u')
      queue = ref([])


    def next()
      # Select two random files
        file_1 = list.hd(!playlist_1)
        file_2 = list.hd(!playlist_2)
    
      # Ensure uniqueness of the two random files
       while string.sub(file_1, start=0, length=string.length(file_1)-10) == string.sub(file_2, start=0, 
                ength=string.length(file_2)-10) do
            file_1 = list.hd(!playlist_1)
            file_2 = list.hd(!playlist_2)
      end

      # Queue of the files to be played n times
        queue := [file_1, file_2, file_1, file_2, file_1, file_2, file_1, file_2]
   end

   request.dynamic(next)
end

# final playlist to be used
hmdecade = decade_playlist()
```

Of course the above code is not working and is more or less to describe the approach to the issue.

**Reply by vitoyucepi** [reply]
Hi,
I don't fully understand your concept of the rotation, so can you provide some test cases?
First of all, I'd like to see the content of the playlist, so I can design a function, that will compare and exclude all non-unique files.
Secondly, it would be great if you could provide the desired rotation generated from the playlists.
And the last one. As I asked earlier, do you have programming skills, for example, in python, because it'll be easier for me to convert the existing script into liquidsoap?

Also, you can use my code and capture the states of the playlists and queues by the `load_playlist_files` tag.
To do that, set `settings.log.level.set(5)`.

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

Here is a simple snipped in Python that would produce the desired results. Of course the numbers in the playlists represent the audio files.

```
import random

# assume two playlists with 5 files each
playlist_1 = [1,2,3,4,5]
playlist_2 = [1,2,3,4,5]
n = 3

def print_rotated_numbers():
    file_1 = random.choice(playlist_1)
    file_2 = random.choice(playlist_2)

    # ensure that the files for each rotation are unique. This will be modified to follow the file naming of the mp3s. 
    while file_1 == file_2:
        file_1 = random.choice(playlist_2)

    for i in range(n):
        print(file_1, file_2)

print_rotated_numbers()
```

Thanks once more!

**Reply by pavlos2** [reply]
Hi @vitoyucepi, can you please still help with the code?

**Reply by vitoyucepi** [reply]
Hi @pavlos2, 

```ruby
def load_playlist_files(~randomize = false, path)
  id = string.id("load_playlist_files")
  let log.info = fun(~label=id, message) -> log.info(label=label, message)
  let log.debug = fun(~label=id, message) -> log.debug(label=label, message)
  let log.important = fun(~label=id, message) -> log.important(label=label, message)

  files = {
    # Source id
    id = id,
    # Playlist path
    path = path,
    # List of files in the playlist
    list = ref([])
  }

  # Is the playlist empty
  let files.is_empty = fun() -> begin
    list.is_empty(!(files.list))
  end

  # Load files to the playlist
  let files.reload = fun() -> begin
    log.info("Reload full playlist")
    files.list := playlist.files(files.path)
    if randomize then
      files.list := list.shuffle(!(files.list))
    end
    
    log.debug("randomize=#{randomize}")
    log.debug("files.list=#{!(files.list)}")
  end
  
  # Get next track in the playlist
  let files.next = fun() -> begin
    log.info("Get next file in the playlist")
    if files.is_empty() then
      log.info("The playlist is empty")
      files.reload()
    end
    
    if files.is_empty() then
      log.important("The playlist can't produce next file")
      null()
    else
      # Get the first element of the list
      next_file_path = list.hd(!(files.list))
      files.list := list.tl(!(files.list))
      log.info("The next file in the playlist is: #{next_file_path}")
      log.debug("files.list=#{!(files.list)}")
      
      next_file_path
    end
  end

  files
end

# Generate rotation
def decade_playlist(~repetitions = 3)

  playlist_1 = load_playlist_files(randomize = true, 'playlist_1.m3u')
  playlist_2 = load_playlist_files(randomize = true, 'playlist_2.m3u')
  queue = ref([])

  # Get next file in queue
  def next()
    if list.is_empty(!queue) then
      playlist_1_next = playlist_1.next()
      playlist_2_next = ref(playlist_2.next())
      while playlist_1_next == !playlist_2_next do
        playlist_2_next := playlist_2.next()
      end
      for i = 1 to repetitions do
         ignore(i)
         queue := list.add(playlist_1_next, !queue)
         queue := list.add(!playlist_2_next, !queue)
      end
    end

    next_file_path = list.hd(!queue)
    queue := list.tl(!queue)

    if null.defined(next_file_path) then
      request.create(null.get(next_file_path))
    else
      null()
    end
  end
  
  request.dynamic(next)
end

# final playlist to be used
output.dummy(fallible = true, decade_playlist())

```

**Reply by pavlos2** [reply]
thanks again @vitoyucepi! I am checking again your code, I needed to change the while loop, from

`while playlist_1_next == !playlist_2_next do`
to
`while string.sub(playlist_1_next, start=0, length=string.length(playlist_1_next)-10) == string.sub(playlist_2_next, start=0, length=string.length(playlist_2_next)-10) do `

however, I get the error about difference types of variables, confused because the involved variables are strings.

**Reply by vitoyucepi** [reply]
Replace `playlist_2_next` with `!playlist_2_next` everywhere.

Or create a function
```ruby
def is_match(name_1, name_2)
  length_1 = string.length(name_1)
  length_2 = string.length(name_2)
  string.sub(name_1, start=0, length=length_1-10) == string.sub(name_2, start=0, length=length_2-10)
end
```
and call it like this
```ruby
while is_match(playlist_1_next, !playlist_2_next) do
  playlist_2_next := playlist_2.next()
end
```

**Reply by pavlos2** [reply]
Hi @vitoyucepi. I have no words to thank you for the support! Tried both suggestions, again and again. I take the same error, the variable should be string.

**Reply by vitoyucepi** [reply]
Hi @pavlos2,
Could you copy-paste the error text?

**Reply by pavlos2** [reply]
Hi @vitoyucepi
I use the script with Azuracast. When I typed !playlist_2_next everywhere I get:

LogicException at /var/azuracast/www/src/Radio/Backend/Liquidsoap.php L246: At line 366, char 70-85: Error 5: this value has type _? (inferred at line 338, char 6-12) but it should be a subtype of string

```
....
336       if files.is_empty() then
337         log.important("The playlist can't produce next file")
338         null()
339       else
....
```

And for the function I get:

LogicException at /var/azuracast/www/src/Radio/Backend/Liquidsoap.php L246: At line 372, char 15-30: Error 5: this value has type _? (inferred at line 338, char 6-12) but it should be a subtype of the type of the value at line 366, char 13-19 string (inferred at line 367, char 27-33)

```
338      null()
...
367     playlist_2_next := playlist_2.next()
...
372      queue := list.add(playlist_1_next, !queue)
```

**Reply by pavlos2** [reply]
Hi @vitoyucepi 
I managed to fix the error as:

 #Define null string
   let null_str = ""

Then, at line 338 I replaced  `null()` with the variable `null_str` so that to get an empty string, when there are no files in the playlist.

Now that I have the full code working the given results are not the expected ones. The final playlist outputs:

....
track_1  track_2
track_3  track_2
track_3  track_2

track_3  track_4
track_2  track_4
track_2  track_4

track_2  track_3
track_4  track_3
track_4  track_3

track_4  track_1
track_2  track_1
track_2  track_1

track_2  track_3
track_4  track_3
track_4  track_3
...

There seems to be some lag in playlist_1 files.

**Reply by vitoyucepi** [reply]
I forgot about that
```ruby
def is_match(name_1, name_2)
  if name_1 == null() or name_2 == null() then
    false
  else
    name_1 = null.get(name_1)
    name_2 = null.get(name_2)
    length_1 = string.length(name_1)
    length_2 = string.length(name_2)
    string.sub(name_1, start=0, length=length_1-10) == string.sub(name_2, start=0, length=length_2-10)
  end
end
```
maybe this will work for you
```ruby
def is_match(name_1, name_2)
  name_1 = null.get(default="", name_1)
  name_2 = null.get(default="", name_2)
  length_1 = string.length(name_1)
  length_2 = string.length(name_2)
  string.sub(name_1, start=0, length=length_1-10) == string.sub(name_2, start=0, length=length_2-10)
end
```

**Reply by pavlos2** [reply]
Hi @vitoyucepi, I tried the above scripts, unfortunately the issue is not fixed. It is getting complicated...

**Reply by vitoyucepi** [reply]
This is the code of the whole script for liquidsoap *2.1.3*.
```ruby
def load_playlist_files(~randomize = false, path)
  id = string.id("load_playlist_files")
  let log.info = fun(~label=id, message) -> log.info(label=label, message)
  let log.debug = fun(~label=id, message) -> log.debug(label=label, message)
  let log.important = fun(~label=id, message) -> log.important(label=label, message)

  files = {
    # Source id
    id = id,
    # Playlist path
    path = path,
    # List of files in the playlist
    list = ref([])
  }

  # Is the playlist empty
  let files.is_empty = fun() -> begin
    list.is_empty(!(files.list))
  end

  # Load files to the playlist
  let files.reload = fun() -> begin
    log.info("Reload full playlist")
    files.list := playlist.files(files.path)
    if randomize then
      files.list := list.shuffle(!(files.list))
    end
    
    log.debug("randomize=#{randomize}")
    log.debug("files.list=#{!(files.list)}")
  end
  
  # Get the next track in the playlist
  let files.next = fun() -> begin
    log.info("Get next file in the playlist")
    if files.is_empty() then
      log.info("The playlist is empty")
      files.reload()
    end
    
    if files.is_empty() then
      log.important("The playlist can't produce next file")
      null()
    else
      # Get the first element of the list
      next_file_path = list.hd(!(files.list))
      files.list := list.tl(!(files.list))
      log.info("The next file in the playlist is: #{next_file_path}")
      log.debug("files.list=#{!(files.list)}")
      
      next_file_path
    end
  end

  files
end

# Generate rotation
def decade_playlist(~repetitions = 3)

  playlist_1 = load_playlist_files(randomize = true, 'playlist_1.m3u')
  playlist_2 = load_playlist_files(randomize = true, 'playlist_2.m3u')
  queue = ref([])

  def is_match(name_1, name_2)
    if name_1 == null() or name_2 == null() then
      false
    else
      name_1 = null.get(name_1)
      name_2 = null.get(name_2)
      length_1 = string.length(name_1)
      length_2 = string.length(name_2)
      string.sub(name_1, start=0, length=length_1-10) == string.sub(name_2, start=0, length=length_2-10)
    end
  end
  # Get the next file in the queue
  def next()
    if list.is_empty(!queue) then
      playlist_1_next = playlist_1.next()
      playlist_2_next = ref(playlist_2.next())
      while is_match(playlist_1_next, !playlist_2_next) do
        playlist_2_next := playlist_2.next()
      end
      for i = 1 to repetitions do
         ignore(i)
         queue := list.add(!playlist_2_next, !queue)
         queue := list.add(playlist_1_next, !queue)
      end
    end

    next_file_path = list.hd(!queue)
    queue := list.tl(!queue)

    if null.defined(next_file_path) then
      request.create(null.get(next_file_path))
    else
      null()
    end
  end
  
  request.dynamic(id="decade_playlist", next)
end

output.dummy(fallible = true, decade_playlist())
```

Here are my playlists
- `playlist_1.m3u`
  ```
  playlist_1/track_4.ogg
  playlist_1/track_3.ogg
  playlist_1/track_2.ogg
  playlist_1/track_1.ogg
  ```
- `playlist_2.m3u`
  ```
  playlist_2/track_4.ogg
  playlist_2/track_3.ogg
  playlist_2/track_2.ogg
  playlist_2/track_1.ogg
  ```

The playback log looks like this
```
2023/02/27 22:48:25 [decade_playlist:3] Prepared "./playlist_1/track_2.ogg" (RID 2).
2023/02/27 22:48:30 [decade_playlist:3] Prepared "./playlist_2/track_4.ogg" (RID 3).
2023/02/27 22:48:35 [decade_playlist:3] Prepared "./playlist_1/track_2.ogg" (RID 4).
2023/02/27 22:48:40 [decade_playlist:3] Prepared "./playlist_2/track_4.ogg" (RID 5).
2023/02/27 22:48:45 [decade_playlist:3] Prepared "./playlist_1/track_2.ogg" (RID 6).
2023/02/27 22:48:50 [decade_playlist:3] Prepared "./playlist_2/track_4.ogg" (RID 7).

2023/02/27 22:48:55 [decade_playlist:3] Prepared "./playlist_1/track_1.ogg" (RID 8).
2023/02/27 22:49:00 [decade_playlist:3] Prepared "./playlist_2/track_1.ogg" (RID 9).
2023/02/27 22:49:05 [decade_playlist:3] Prepared "./playlist_1/track_1.ogg" (RID 10).
2023/02/27 22:49:10 [decade_playlist:3] Prepared "./playlist_2/track_1.ogg" (RID 11).
2023/02/27 22:49:15 [decade_playlist:3] Prepared "./playlist_1/track_1.ogg" (RID 12).
2023/02/27 22:49:20 [decade_playlist:3] Prepared "./playlist_2/track_1.ogg" (RID 13).
```

**Reply by pavlos2** [reply]
Hi @vitoyucepi,
Indeed, the first iteration [22:48:25 - 22:48:50] of the above logs depicts the expected results, when the function `decade_playlist()` is called. The second iteration will play 6 times the same file, why?

**Reply by vitoyucepi** [reply]
The files are not the same. They are from different playlists.

**Reply by pavlos2** [reply]
Indeed, you are right! Somehow, I cannot output the same results, the script is run within Azuracast, I have to dig more to see what is wrong. I really appreciate your huge support, are you a freelancer programmer?

**Reply by vitoyucepi** [reply]
@pavlos2
> are you a freelancer programmer?

Not at all. Just trying to relieve some load from the maintainers of this repo.
I know the most common functions of the liquidsoap, but I don't know ocaml.
I'm not even running an instance of the liquidsoap at the moment.

As for the programming, I prefer to deal with Linux servers, containers, and deployments.

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

So far I run many tests trying to solve the issue, to prevent playing back-to-back tracks with the same voice. I used gpt to analyze the code many times, it is possible that it does not handles edge cases correctly. When the decade_playlist function repeats the pairs of tracks, according to the repetitions parameter, it doesn't check if the same voice appears at the end of one repetition and the beginning of the next. This could lead to the same voice playing back-to-back. Your thoughts?

**Reply by vitoyucepi** [reply]
Hi @pavlos2,

It's possible to utilize the `playlog` function [page 75](http://www.liquidsoap.info/book/book.pdf).
Also, it's possible to add custom playlog tracking to the `load_playlist_files` function.

There's also a potential bug here.
```ruby
while is_match(playlist_1_next, !playlist_2_next) do
  playlist_2_next := playlist_2.next()
end
```

**Reply by pavlos2** [reply]
Hi @vitoyucepi,

Thanks again for your help. Could you be more specific about the bug?

The more we work on this code I suspect there may be timing issues too, because the duration of audio files used for this particular project is less than half a minute.

Also, I find that too much code is needed to implement: **A custom file rotation within two playlists, playlist_1 and playlist_2. It alternates between the tracks of these two playlists, ensuring that back-to-back tracks do not have the same voice, and repeats this process with a specified number of repetitions before moving to the next pair of tracks from the playlists.**
